deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py

```diff
@@ -0,0 +1,146 @@
+                  'hidden_layer_width': self.hidden_layer_width,
+                  'initialization': self.hidden_layer_init,
+                  'hidden_layer_activation': self.hidden_layer_activation,
+                  'final_activation': self.final_activation}
+        config.update(base_config)
+        return config
+
+    def get_output_shape_for(self, input_shapes):
+        # pylint: disable=unused-argument
+        return (input_shapes[0][0], 1)
+
+    def build(self, input_shape):
+        super(SlotSimilarityTupleMatcher, self).build(input_shape)
+
+        # Add the weights for the hidden layers.
+        hidden_layer_input_dim = input_shape[0][1]
+        for i in range(self.num_hidden_layers):
+            hidden_layer = self.add_weight(shape=(hidden_layer_input_dim, self.hidden_layer_width),
+                                           initializer=initializations.get(self.hidden_layer_init),
+                                           name='%s_hiddenlayer_%d' % (self.name, i))
+            self.hidden_layer_weights.append(hidden_layer)
+            hidden_layer_input_dim = self.hidden_layer_width
+        # Add the weights for the final layer.
+        self.score_layer = self.add_weight(shape=(self.hidden_layer_width, 1),
+                                           initializer=initializations.get(self.hidden_layer_init),
+                                           name='%s_score' % self.name)
+
+    @overrides
+    def compute_mask(self, input, input_mask=None):  # pylint: disable=unused-argument,redefined-builtin
+        # Here, we want to see if either of the inputs is all padding (i.e. the mask would be all 0s).
+        # If so, then the whole tuple_match should be masked, so we would return a 0, otherwise we
+        # return a 1.  As such, the shape of the returned mask is (batch size, 1).
+        if input_mask is None or input_mask == [None, None]:
+            return None
+        # Each of the two masks in input_mask are of shape: (batch size, num_slots)
+        mask1, mask2 = input_mask
+        mask1 = K.cast(K.any(mask1), 'int32')
+        mask2 = K.cast(K.any(mask2), 'int32')
+        return (mask1 + mask2) >= 2
+
+    def call(self, x, mask=None):
+        tuple1_input, tuple2_input = x      # tuple1 shape: (batch size, num_slots, embed_dimension)
+                                            # tuple2 shape: (batch size, num_slots, embed_dimension)
+        # Check that the tuples have the same number of slots.
+        assert K.int_shape(tuple1_input)[1] == K.int_shape(tuple2_input)[1]
+
+        # Calculate the cosine similarities.
+        # shape: (batch size, num_slots)
+        similarities = self.similarity_function.compute_similarity(tuple1_input, tuple2_input)
+
+        # Remove any similarities if one of the corresponding slots was all padding.
+        # TODO(becky): need this to take a mask for each input of None or shape: (batch size, num_slots)
```

Comment:
I think this TODO can be removed?  Or is there something else that needs to be done?